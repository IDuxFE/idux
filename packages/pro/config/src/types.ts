/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ProLocale } from '@idux/pro/locales'
import type { ProTableColumnIndexable } from '@idux/pro/table'

export interface ProGlobalConfig {
  common: ProCommonConfig
  locale: ProLocale

  table: ProTableConfig
  tree: ProTreeConfig
}

export type ProGlobalConfigKey = keyof ProGlobalConfig

export interface ProCommonConfig {
  prefixCls: string
}

export interface ProTableConfig {
  columnIndexable: Omit<ProTableColumnIndexable, 'type'>
}

export interface ProTreeConfig {
  clearIcon: string
  collapseIcon: [string, string]
}
