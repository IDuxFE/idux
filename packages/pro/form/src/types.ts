/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import type { AbstractControl, AsyncValidatorFn, ValidatorFn } from '@idux/cdk/forms'
import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray } from '@idux/cdk/utils'
import type { FormItemProps, FormProps } from '@idux/components/form'
import type { RowProps } from '@idux/components/grid'
import type { Options as AjvOptions } from 'ajv'
import type { Component, DefineComponent, HTMLAttributes, PropType, VNodeChild } from 'vue'

export const proFormProps = {
  ajvOptions: Object as PropType<AjvOptions>,
  autoId: { type: [Boolean, String] as PropType<boolean | string>, default: undefined },
  autoLabelFor: { type: Boolean, default: undefined },
  fields: Object as PropType<ProFormFieldsSchema>,
  ignoreKeywords: Array as PropType<string[]>,
  schema: Object as PropType<ProFormJsonSchema>,
  schemaFormatter: Function as PropType<ProFormSchemaFormatter>,

  //events
  onSubmit: [Function, Array] as PropType<MaybeArray<(formData: any) => void>>,
} as const

export type ProFormProps = ExtractInnerPropTypes<typeof proFormProps>
export type ProFormPublicProps = ExtractPublicPropTypes<typeof proFormProps>
export type ProFormComponent = DefineComponent<
  Omit<HTMLAttributes, keyof ProFormPublicProps> & ProFormPublicProps & FormProps
>
export type ProFormInstance = InstanceType<DefineComponent<ProFormProps>>

type ArrayElement<A> = A extends (infer T)[] ? T : never

export type ProFormSchemaFormatter = (
  fields: ProFormFieldsSchema | undefined,
  schema: ProFormJsonSchema | undefined,
) => { fields: ProFormFieldsSchema; schema: ProFormJsonSchema }

// custom render(slots)
export interface ProFormItemCustoms {
  customControl?: string | ((props: { control: AbstractControl; [key: string]: any }) => VNodeChild)
  customControlTooltip?: string | (() => VNodeChild)
  customDescription?: string | (() => VNodeChild)
  customLabel?: string | (() => VNodeChild)
  customLabelTooltip?: string | (() => VNodeChild)
  customMessage?: string | (() => VNodeChild)
}

export interface ProFormFieldsSchema<T = any> extends ProFormItemCustoms {
  formProps?: FormProps
  formItemProps?: FormItemProps & RowProps
  component?: string | Component
  componentProps?: any

  default?: T

  disabled?: boolean
  name?: string
  example?: string
  trigger?: 'change' | 'blur' | 'submit'
  validators?: ValidatorFn | ValidatorFn[]
  asyncValidators?: AsyncValidatorFn | AsyncValidatorFn[]

  order?: string[]
  type?: 'object' | 'array'

  // show?: boolean | ((control: AbstractControl) => boolean)
  // if?: boolean | ((control: AbstractControl) => boolean)

  properties?: { readonly [K in keyof T]: ProFormFieldsSchema<T[K]> }
  items?: ProFormFieldsSchema<ArrayElement<T>>
}

export type ProFormJsonSchema<T = any> =
  | ObjectKeywords<T>
  | ArrayKeywords<T>
  | StringKeywords<T>
  | NumberKeywords<T>
  | BooleanKeywords<T>

export interface CommonKeywords<T> {
  title?: string
  description?: string
  const?: any
  enum?: Readonly<(T | null)[]>
  default?: T | null
}

export interface ObjectKeywords<T> extends CommonKeywords<T> {
  type?: 'object'
  properties: { readonly [K in keyof T]: ProFormJsonSchema<T[K]> }
  required?: (keyof T)[]
}

export interface ArrayKeywords<T> extends CommonKeywords<T> {
  type: 'array'
  items?: ProFormJsonSchema<ArrayElement<T>>
  minItems?: number
  maxItems?: number
}

export interface StringKeywords<T> extends CommonKeywords<T> {
  type: 'string'
  minLength?: number
  maxLength?: number
  pattern?: string
  format?: string
}

export interface NumberKeywords<T> extends CommonKeywords<T> {
  type: 'number' | 'integer'
  minimum?: number
  maximum?: number
  format?: string
}

export interface BooleanKeywords<T> extends CommonKeywords<T> {
  type: 'boolean'
}
