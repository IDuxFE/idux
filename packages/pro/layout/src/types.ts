/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { IxInnerPropTypes, IxPublicPropTypes, VKey } from '@idux/cdk/utils'
import type { LayoutSiderProps } from '@idux/components/layout'
import type { MenuClickOptions, MenuData, MenuProps, MenuTheme } from '@idux/components/menu'
import type { DefineComponent, HTMLAttributes } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

export interface SiderHover {
  delay: number
}

export interface SiderHoverCtrl extends SiderHover {
  enable: boolean
}

export const proLayoutProps = {
  activeKey: IxPropTypes.oneOfType<VKey>([String, Number, Symbol]),
  collapsed: IxPropTypes.bool,
  siderHover: IxPropTypes.oneOfType([IxPropTypes.bool, IxPropTypes.object<SiderHover>()]),
  compress: IxPropTypes.bool.def(true),
  fixed: IxPropTypes.oneOfType([Boolean, IxPropTypes.object<{ header: boolean; sider: boolean }>()]).def(false),
  menus: IxPropTypes.array<MenuData>().def(() => []),
  sider: IxPropTypes.object<LayoutSiderProps>(),
  siderMenu: IxPropTypes.object<MenuProps>(),
  theme: IxPropTypes.oneOfType([
    IxPropTypes.oneOf(['light', 'dark']),
    IxPropTypes.object<{ header: MenuTheme; sider: MenuTheme }>(),
  ]).def('light'),
  type: IxPropTypes.oneOf<ProLayoutType>(['header', 'sider', 'both', 'mixin']).def('mixin'),

  // event
  'onUpdate:activeKey': IxPropTypes.emit<(activeKey: VKey | null) => void>(),
  'onUpdate:collapsed': IxPropTypes.emit<(collapsed: boolean) => void>(),
  onMenuClick: IxPropTypes.emit<(options: MenuClickOptions) => void>(),
}

export type ProLayoutProps = IxInnerPropTypes<typeof proLayoutProps>
export type ProLayoutPublicProps = IxPublicPropTypes<typeof proLayoutProps>
export type ProLayoutComponent = DefineComponent<
  Omit<HTMLAttributes, keyof ProLayoutPublicProps> & ProLayoutPublicProps
>
export type ProLayoutInstance = InstanceType<DefineComponent<ProLayoutProps>>

export type ProLayoutFixed = boolean | { header: boolean; sider: boolean }
export type ProLayoutType = 'header' | 'sider' | 'mixin' | 'both'

export const proLayoutSiderTriggerProps = {
  icon: IxPropTypes.arrayOf(IxPropTypes.oneOfType([String, IxPropTypes.vNode])),
}

export type ProLayoutSiderTriggerProps = IxInnerPropTypes<typeof proLayoutSiderTriggerProps>
export type ProLayoutSiderTriggerPublicProps = IxPublicPropTypes<typeof proLayoutSiderTriggerProps>
export type ProLayoutSiderTriggerComponent = DefineComponent<
  Omit<HTMLAttributes, keyof ProLayoutSiderTriggerPublicProps> & ProLayoutSiderTriggerPublicProps
>
export type ProLayoutSiderTriggerInstance = InstanceType<DefineComponent<ProLayoutSiderTriggerProps>>
