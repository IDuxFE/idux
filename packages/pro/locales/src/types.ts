/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

export interface ProTableLocale {
  layout: {
    title: string
    sm: string
    md: string
    lg: string
    placeholder: string
    all: string
    reset: string
    indexable: string
    expandable: string
    selectable: string
    startPin: string
    endPin: string
    noPin: string
    startPinTitle: string
    endPinTitle: string
    noPinTitle: string
  }
}

export interface ProTreeLocale {
  expandAll: string
  collapseAll: string
}

export interface ProSearchLocale {
  keyword: string
  ok: string
  cancel: string
  selectAll: string
  placeholder: string
  switchToTimePanel: string
  switchToDatePanel: string
}

export interface ProLocale {
  type: ProLocaleType

  table: ProTableLocale
  tree: ProTreeLocale
  search: ProSearchLocale
}

export type ProLocaleType = 'zh-CN' | 'en-US'
