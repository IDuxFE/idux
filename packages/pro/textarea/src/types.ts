/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray } from '@idux/cdk/utils'
import type { TextareaResize } from '@idux/components/textarea'
import type { DefineComponent, HTMLAttributes, PropType } from 'vue'

export interface TextareaError {
  index: number
  message?: string
}

import { ɵCommonProps } from '@idux/components/input'

export const proTextareaProps = {
  ...ɵCommonProps,
  computeCount: { type: Function as PropType<(value: string) => string>, default: undefined },
  errors: Array as PropType<TextareaError[]>,
  maxCount: { type: [Number, String] as PropType<number | string>, default: undefined },
  placeholder: String,
  resize: { type: String as PropType<TextareaResize>, default: undefined },
  showCount: { type: Boolean, default: undefined },

  onKeyDown: [Function, Array] as PropType<MaybeArray<(evt: KeyboardEvent) => void>>,
} as const

export type ProTextareaProps = ExtractInnerPropTypes<typeof proTextareaProps>
export type ProTextareaPublicProps = ExtractPublicPropTypes<typeof proTextareaProps>
export type ProTextareaComponent = DefineComponent<
  Omit<HTMLAttributes, keyof ProTextareaPublicProps> & ProTextareaPublicProps
>
export type ProTextareaInstance = InstanceType<DefineComponent<ProTextareaProps>>

export const errorLineProps = {
  message: String,
  visible: Boolean,
} as const

export type ErrorLineProps = ExtractInnerPropTypes<typeof errorLineProps>

export type { TextareaResize } from '@idux/components/textarea'
