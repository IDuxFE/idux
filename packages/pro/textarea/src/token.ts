/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ErrorLinesContext } from './composables/useErrorLines'
import type { ProTextareaProps } from './types'
import type { ɵBoxSizingData } from '@idux/components/textarea'
import type { ComputedRef, InjectionKey, Ref } from 'vue'

export interface ProTextareaContext extends ErrorLinesContext {
  props: ProTextareaProps
  boxSizingData: ComputedRef<ɵBoxSizingData>
  lineHeight: ComputedRef<number>
  mergedPrefixCls: ComputedRef<string>
  rowCounts: ComputedRef<number[]>
  textareaRef: Ref<HTMLTextAreaElement | undefined>
  handleInput: (evt: Event) => void
  handleCompositionStart: (evt: CompositionEvent) => void
  handleCompositionEnd: (evt: CompositionEvent) => void
}

export const proTextareaContext: InjectionKey<ProTextareaContext> = Symbol('proTextareaContext')
