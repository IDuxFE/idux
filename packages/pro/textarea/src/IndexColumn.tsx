/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, inject, normalizeClass, normalizeStyle } from 'vue'

import { proTextareaContext } from './token'

export default defineComponent({
  setup() {
    const { props, mergedPrefixCls, rowCounts, lineHeight, boxSizingData, setvisibleErrIndex } =
      inject(proTextareaContext)!

    const mergedRowCnts = computed(() => (rowCounts.value.length > 0 ? rowCounts.value : [0]))
    const columnStyle = computed(() =>
      normalizeStyle({
        paddingTop: `${boxSizingData.value?.paddingTop ?? 0}px`,
        paddingBottom: `${boxSizingData.value?.paddingBottom ?? 0}px`,
      }),
    )

    return () => {
      const prefixCls = `${mergedPrefixCls.value}-index-column`

      return (
        <div class={prefixCls} style={columnStyle.value}>
          {mergedRowCnts.value.map((cnt, index) => {
            const cellStyle = normalizeStyle({
              height: `${cnt * lineHeight.value}px`,
            })

            const errorIdx = props.errors?.find(error => error.index === index)?.index ?? -1
            const cellClass = normalizeClass({
              [`${prefixCls}-cell`]: true,
              [`${prefixCls}-cell-error`]: errorIdx > -1,
            })

            const handleCellMouseEnter = () => {
              setvisibleErrIndex(errorIdx)
            }
            const handleCellMouseLeave = () => {
              setvisibleErrIndex(-1)
            }

            return (
              <div
                class={cellClass}
                style={cellStyle}
                onMouseenter={handleCellMouseEnter}
                onMouseleave={handleCellMouseLeave}
              >
                {index + 1}
              </div>
            )
          })}
        </div>
      )
    }
  },
})
