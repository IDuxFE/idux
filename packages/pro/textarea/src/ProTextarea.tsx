/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import {
  type Ref,
  computed,
  defineComponent,
  normalizeClass,
  normalizeStyle,
  onMounted,
  provide,
  ref,
  toRef,
  watch,
} from 'vue'

import { callEmit } from '@idux/cdk/utils'
import { IxIcon } from '@idux/components/icon'
import { ɵUseInput } from '@idux/components/input'
import { ɵUseAutoRows } from '@idux/components/textarea'
import { type ProTextareaConfig, useGlobalConfig } from '@idux/pro/config'

import IndexColumn from './IndexColumn'
import { useErrorLines } from './composables/useErrorLines'
import { useRowCounts } from './composables/useRowsCounts'
import Content from './content/Content'
import { proTextareaContext } from './token'
import { ProTextareaProps, proTextareaProps } from './types'

export default defineComponent({
  name: 'IxProTextarea',
  props: proTextareaProps,
  setup(props, { slots, expose }) {
    const common = useGlobalConfig('common')
    const config = useGlobalConfig('textarea')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-textarea`)

    const {
      elementRef,
      accessor,
      isFocused,
      isComposing,
      clearable,
      clearIcon,
      clearVisible,
      handleInput,
      handleClear,
      handleCompositionStart,
      handleCompositionEnd,
      focus,
      blur,
      syncValue,
    } = ɵUseInput<HTMLTextAreaElement>(props, config)
    const valueRef = toRef(accessor, 'value')

    const { lineHeight, boxSizingData, resizeToFitContent } = ɵUseAutoRows(elementRef, ref(true))
    const rowCounts = useRowCounts(elementRef, valueRef, lineHeight, boxSizingData)
    const errorLinesContext = useErrorLines(props, lineHeight, boxSizingData)
    const dataCount = useDataCount(props, config, valueRef)

    const _handleInput = (evt: Event) => {
      if (isComposing.value) {
        resizeToFitContent()
      }

      callEmit(handleInput, evt)
    }

    onMounted(() => {
      syncValue()
      watch(() => accessor.value, resizeToFitContent, { immediate: true })
    })

    expose({ focus, blur })

    provide(proTextareaContext, {
      mergedPrefixCls,
      props,
      boxSizingData,
      lineHeight,
      rowCounts,
      textareaRef: elementRef,
      ...errorLinesContext,
      handleInput: _handleInput,
      handleCompositionStart,
      handleCompositionEnd,
    })

    const style = computed(() => normalizeStyle({ resize: props.resize ?? config.resize }))
    const classes = computed(() => {
      const prefixCls = mergedPrefixCls.value
      const size = props.size ?? config.size

      return normalizeClass({
        [prefixCls]: true,
        [`${prefixCls}-disabled`]: accessor.disabled,
        [`${prefixCls}-focused`]: isFocused.value,
        [`${prefixCls}-clearable`]: clearable.value,
        [`${prefixCls}-${size}`]: !!size,
      })
    })

    const handleScrollHolderMouseDown = (evt: MouseEvent) => {
      if (evt.target === elementRef.value) {
        return
      }

      if (!isFocused.value) {
        focus()
      }

      evt.preventDefault()
    }

    return () => {
      const prefixCls = mergedPrefixCls.value

      return (
        <span class={classes.value} style={style.value}>
          <div class={`${prefixCls}-index-column-bg`}></div>
          <div class={`${prefixCls}-scroll-holder`} onMousedown={handleScrollHolderMouseDown}>
            <div class={`${prefixCls}-inner`}>
              <IndexColumn />
              <Content />
            </div>
          </div>
          {clearable.value && (
            <span
              class={normalizeClass([`${prefixCls}-clear`, clearVisible.value ? 'visible' : ''])}
              onClick={handleClear}
            >
              {slots.clearIcon ? slots.clearIcon() : <IxIcon name={clearIcon.value}></IxIcon>}
            </span>
          )}
          {dataCount.value && <span class={`${prefixCls}-count`}>{dataCount.value}</span>}
        </span>
      )
    }
  },
})

function useDataCount(props: ProTextareaProps, config: ProTextareaConfig, valueRef: Ref<string | undefined>) {
  return computed(() => {
    const showCount = props.showCount ?? config.showCount
    const computeCount = props.computeCount ?? config.computeCount
    const maxCount = props.maxCount ?? config.maxCount

    let dataCount = ''
    if (showCount) {
      const value = valueRef.value ?? ''
      dataCount = `${value.length}`
      if (computeCount) {
        dataCount = computeCount(value)
      } else if (maxCount) {
        dataCount += ' / ' + maxCount
      }
    }
    return dataCount
  })
}

// function useTexarea(
//   props: ProTextareaProps,
//   config: ProTextareaConfig,
//   textareaCompRef: Ref<TextareaInstance | undefined>,
//   valueRef: Ref<string | undefined>,
//   setValue: (value: string | undefined) => void,
// ): {
//   textareaRef: Ref<HTMLTextAreaElement | undefined>
//   lineHeight: ComputedRef<number>
//   boxSizingData: ComputedRef<ɵBoxSizingData>
//   isFocused: ComputedRef<boolean>
//   textareaCompProps: ComputedRef<TextareaProps>
// } {
//   const textareaRef = ref<HTMLTextAreaElement | undefined>()
//   const [isFocused, setIsFocused] = useState<boolean>(false)

//   watch(textareaCompRef, textareaComp => {
//     nextTick(() => {
//       textareaRef.value = textareaComp?.getTextareaElement()
//     })
//   })

//   const lineHeight = ɵUseLineHeight(textareaRef)
//   const boxSizingData = useTextareaBoxSizingData(textareaRef)

//   const textareaCompProps = computed<TextareaProps>(() => {
//     const {
//       control,
//       clearable = config.clearable,
//       clearIcon = config.clearIcon,
//       disabled,
//       readonly,
//       size = config.size,
//       trim,

//       onChange,
//       onClear,
//       onCompositionStart,
//       onCompositionEnd,
//       onInput,
//       onFocus,
//       onBlur,
//     } = props

//     const handleTextareaCompFocus = (evt: FocusEvent) => {
//       callEmit(onFocus, evt)
//       setIsFocused(true)
//     }
//     const handleTextareaCompBlur = (evt: FocusEvent) => {
//       callEmit(onBlur, evt)
//       setIsFocused(false)
//     }

//     return {
//       value: valueRef.value,
//       autoRows: true,
//       borderless: true,
//       showCount: false,
//       control,
//       clearable,
//       clearIcon,
//       disabled,
//       readonly,
//       size,
//       trim,

//       'onUpdate:value': setValue,
//       onChange,
//       onClear,
//       onCompositionStart,
//       onCompositionEnd,
//       onInput,
//       onFocus: handleTextareaCompFocus,
//       onBlur: handleTextareaCompBlur,
//     }
//   })

//   return {
//     textareaRef,
//     lineHeight,
//     boxSizingData,
//     isFocused,
//     textareaCompProps,
//   }
// }
