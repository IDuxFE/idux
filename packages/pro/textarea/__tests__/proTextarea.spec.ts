import { MountingOptions, mount } from '@vue/test-utils'

import { renderWork } from '@tests'

import ProTextarea from '../src/ProTextarea'
import { ProTextareaProps } from '../src/types'

describe.skip('ProTextarea', () => {
  const ProTextareaMount = (options?: MountingOptions<Partial<ProTextareaProps>>) =>
    mount(ProTextarea, { ...(options as MountingOptions<ProTextareaProps>) })

  renderWork<ProTextareaProps>(ProTextarea, {
    props: {},
  })

  test('xxx work', async () => {
    const wrapper = ProTextareaMount({ props: { xxx: 'Xxx' } })

    expect(wrapper.classes()).toContain('ix-Xxx')

    await wrapper.setProps({ xxx: 'Yyy' })

    expect(wrapper.classes()).toContain('ix-Yyy')
  })
})
