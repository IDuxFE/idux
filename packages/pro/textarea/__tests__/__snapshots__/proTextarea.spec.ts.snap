// Vitest Snapshot v1

exports[`ProTextarea > render work 1`] = `"<!---->"`;
